src/coreclr/vm/pendingload.cpp

```diff
@@ -12,241 +12,365 @@
-        PRECONDITION(CheckPointer(pData));
-        PRECONDITION(FindItem(&pData->GetTypeKey()) == NULL);
+        PRECONDITION(HasLock());
+        PRECONDITION(m_dwWaitCount > 0);
     }
-    CONTRACTL_END
-
-    _ASSERTE(m_dwNumBuckets != 0);
-
-    DWORD           dwHash = HashTypeKey(&pData->GetTypeKey());
-    DWORD           dwBucket = dwHash % m_dwNumBuckets;
-    PendingTypeLoadTable::TableEntry * pNewEntry = AllocNewEntry();
-    if (pNewEntry == NULL)
-        return FALSE;
+    CONTRACTL_END;
 
-    // Insert at head of bucket
-    pNewEntry->pNext        = m_pBuckets[dwBucket];
-    pNewEntry->pData        = pData;
-    pNewEntry->dwHashValue  = dwHash;
+    _ASSERTE(m_fLockAcquired);
+    m_Crst.Leave();
+    m_fLockAcquired = FALSE;
+}
 
-    m_pBuckets[dwBucket] = pNewEntry;
+const TypeKey* PendingTypeLoadTable::Entry::GetTypeKey()
+{
+    LIMITED_METHOD_CONTRACT;
+    return &m_typeKey;
+}
 
-    return TRUE;
+void PendingTypeLoadTable::Entry::AddRef()
+{
+    LIMITED_METHOD_CONTRACT;
+    InterlockedIncrement(&m_dwWaitCount);
 }
 
-BOOL PendingTypeLoadTable::DeleteValue(TypeKey *pKey)
+void PendingTypeLoadTable::Entry::Release()
 {
-    CONTRACTL
+    LIMITED_METHOD_CONTRACT;
+    if (InterlockedDecrement(&m_dwWaitCount) == 0)
     {
-        NOTHROW;
-        GC_NOTRIGGER;
-        MODE_ANY;
-        FORBID_FAULT;
-        PRECONDITION(CheckPointer(pKey));
-    }
-    CONTRACTL_END
+        Reset();
 
-    _ASSERTE(m_dwNumBuckets != 0);
+        if (this->m_fIsPreallocated)
+        {
+            // We won't be holding the lock while Releasing, so use a VolatileStore to ensure all writes during Reset are complete.
+            VolatileStore(&m_fIsUnused, true);
+        }
+        else
+        {
+            // Call the derived type with a destructor
+            delete static_cast<DynamicallyAllocatedEntry*>(this);
+        }
+    }
+}
 
-    DWORD           dwHash = HashTypeKey(pKey);
-    DWORD           dwBucket = dwHash % m_dwNumBuckets;
-    PendingTypeLoadTable::TableEntry * pSearch;
-    PendingTypeLoadTable::TableEntry **ppPrev = &m_pBuckets[dwBucket];
+bool PendingTypeLoadTable::Entry::HasWaiters()
+{
+    LIMITED_METHOD_CONTRACT;
+    return m_dwWaitCount > 1;
+}
 
-    for (pSearch = m_pBuckets[dwBucket]; pSearch; pSearch = pSearch->pNext)
+HRESULT PendingTypeLoadTable::Entry::DelayForProgress(TypeHandle* typeHndWithProgress)
+{
+    STANDARD_VM_CONTRACT;
+    HRESULT hr = S_OK;
     {
-        TypeKey entryTypeKey = pSearch->pData->GetTypeKey();
-        if (pSearch->dwHashValue == dwHash && TypeKey::Equals(pKey, &entryTypeKey))
+        CrstHolder crstHolder(&m_Crst);
+        _ASSERTE(HasLock());
+        hr = m_hrResult;
+
+        if (SUCCEEDED(hr))
         {
-            *ppPrev = pSearch->pNext;
-            FreeEntry(pSearch);
-            return TRUE;
+            *typeHndWithProgress = m_typeHandle;
         }
-
-        ppPrev = &pSearch->pNext;
     }
 
-    return FALSE;
+    return m_hrResult;
```

Comment:
```suggestion
    return hr;
```

I assume the logic here is we want to return the operation we branched on. I'm being defensive for non-obvious races.